tool/infocert.c

```diff
@@ -0,0 +1,159 @@
+    xtt_ecdsap256_pub_key public = {.data = {0}};
+    xtt_certificate_expiry expiry = {.data = {0}};
+    char common_name[sizeof(xtt_identity_string)] = {0};
+
+    memcpy((char*)expiry.data, (char*) &certificate[113], sizeof(xtt_certificate_expiry)-2);
+    printf("Expiry: ");
+    printf("20%s\n", expiry.data);
+
+    memcpy(common_name, (char*) &certificate[55], sizeof(xtt_identity_string)-1);
+    printf("Common Name: ");
+    printf("%s\n", common_name);
+
+    memcpy((char*)public.data, (char*) &certificate[204], sizeof(xtt_ecdsap256_pub_key));
+    printf("Public key: {");
+    for(unsigned int i = 0; i < sizeof(xtt_ecdsap256_pub_key); i++){
+        printf("%#02X", public.data[i]);
+        if (i < (sizeof(xtt_ecdsap256_pub_key)-1)) {
+            printf(", ");
+        } else {
+            printf("}\n");
+        }
+    }
+
+    return;
+}
+
+static
+void server_cert(unsigned char* certificate){
+    int index = 0;
+    printf("Server ID: ");
+    for(unsigned int i = 0; i < sizeof(xtt_identity_type); i++){
+        if(certificate[i]<0x10)
+            printf("0%x", certificate[i]);
+        else
+            printf("%x", certificate[i]);
+    }
+    printf("\n");
+    index+= sizeof(xtt_identity_type);
+
+    printf("Expiry: ");
+    for(unsigned int i = index; i < index+sizeof(xtt_certificate_expiry); i++){
+        printf("%c", certificate[i]);
+    }
+    printf("\n");
+    index += sizeof(xtt_certificate_expiry);
+
+    printf("Root ID: ");
+    for(unsigned int i = index; i < index + sizeof(xtt_certificate_root_id); i++){
+        if(certificate[i]<0x10)
+            printf("0%x", certificate[i]);
+        else
+            printf("%x", certificate[i]);
+    }
+    printf("\n");
+    index += sizeof(xtt_certificate_root_id);
+
+    printf("Public Key: {");
+    for(unsigned int i = index; i < index+sizeof(xtt_ecdsap256_pub_key); i++){
+        printf("%#02X", certificate[i]);
+        if (i < (index+sizeof(xtt_ecdsap256_pub_key)-1)) {
+            printf(", ");
+        } else {
+            printf("}\n");
+        }
+    }
+
+    return;
+}
+
+static
+void root_cert(unsigned char* certificate){
+    int index = 0;
+    printf("Root ID: ");
+    for(unsigned int i = 0; i < sizeof(xtt_identity_type); i++){
+        if(certificate[i]<0x10)
+            printf("0%x", certificate[i]);
+        else
+            printf("%x", certificate[i]);
+    }
+    printf("\n");
+    index += sizeof(xtt_identity_type);
+
+    printf("Public key: {");
+    for(unsigned int i = index; i < index+sizeof(xtt_ecdsap256_pub_key); i++){
+        if(certificate[i]<0x10)
+            printf("0x0%x", certificate[i]);
+        else
+            printf("0x%x", certificate[i]);
+
+        if (i < (index+sizeof(xtt_ecdsap256_pub_key)-1)) {
+            printf(", ");
+        } else {
+            printf("}\n");
+        }
+    }
+
+    return;
+}
+
+
+void info_cert(const char* x509_file, const char* server_certificate_filename, const char* root_cert_file){
```

Comment:
This structure is a bit odd.

Maybe take a one file name and an enum indicating the type of certificate:

```
enum infocert_type {infocert_type_x509, infocert_type_server, infocert_type_root}; 

void info_cert(enum infocert_type type, const char* filename);
```

In the current form, it looks like `info_cert` will print all three types, when if in fact it just prints the first non-null type.

Also, APIs that expect NULL values are ugly and code smell.  It's almost always better to redesign the API so that no parameters should be NULL.  (That's not always achievable or reasonable in C, but I think it is here.)

Reply:
`void info_cert` should return an error code (`int info_cert`), so that the `xtt` tool can exit with a meaningful exit code.  Right now, it will `exit(0)` even if the file was invalid.